ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/HD/BackingStore/Lockstep.hs

```diff
@@ -0,0 +1,749 @@
+data IOLikeMonad m a where
+  RealIO :: IO a -> IOLikeMonad IO a
+  SimIO  :: IOSim s a -> IOLikeMonad (IOSim s) a
+
+-- | Retrieve the wrapped @'IOLike'@ monad.
+unIOLikeMonad :: IOLikeMonad m a -> m a
+unIOLikeMonad (RealIO x) = x
+unIOLikeMonad (SimIO x)  = x
+
+-- | Create a wrapper @'IOLike'@ monad.
+class IOLikeMonadC m where
+  ioLikeMonad :: m a -> IOLikeMonad m a
+
+instance IOLikeMonadC IO where
+  ioLikeMonad x = RealIO x
+
+instance IOLikeMonadC (IOSim s) where
+  ioLikeMonad x = SimIO x
+
+instance (Functor m, IOLikeMonadC m) => Functor (IOLikeMonad m) where
+  fmap f x = ioLikeMonad $ fmap f (unIOLikeMonad x)
+
+instance (Applicative m, IOLikeMonadC m) =>Applicative (IOLikeMonad m) where
+  x <*> y = ioLikeMonad $ unIOLikeMonad x <*> unIOLikeMonad y
+  pure = ioLikeMonad . pure
+
+instance (Monad m, IOLikeMonadC m) => Monad (IOLikeMonad m) where
+  m >>= fm = ioLikeMonad $ unIOLikeMonad m >>= unIOLikeMonad . fm
+
+-- | Since the tests do not return any types specific to the underlying
+-- @'IOLike'@ monad, @'Realized' ('IOLikeMonad' m)@ behaves just like
+-- @'Realized' 'IO'@.
+type instance Realized (IOLikeMonad m) a = a
+
+{-------------------------------------------------------------------------------
+  @'Values'@ wrapper
+-------------------------------------------------------------------------------}
+
+-- | Wrapper for preventing nonsenical pattern matches.
+--
+-- A logical step is to have the @'BSVHRangeRead'@ and @'BSVHRead'@ actions
+-- declare that the result of the action should be something of type @'vs'@.
+-- However, this means that in theory @'vs'@ could be instantiated to any type
+-- (like @'Handle'@). Consequentially, if we match on a value that is returned
+-- by running an action, we would always have to match on the case where it is a
+-- result of running @'BSVHRangeRead'@ and @'BSVHRead'@ as well, even if the
+-- return type is @'Handle'@, which we don't expect to use as our @vs@ type. As
+-- such, we define this wrapper to prevent having to match on this nonsensical
+-- case.
+newtype Values vs = Values {unValues :: vs}
+  deriving stock (Show, Eq, Ord, Typeable)
+  deriving newtype QC.Arbitrary
+
+{-------------------------------------------------------------------------------
+  Model state
+-------------------------------------------------------------------------------}
+
+data BackingStoreState ks vs d = BackingStoreState {
+    bssMock  :: Mock vs
+  , bssStats :: Stats ks vs d
+  }
+  deriving (Show, Eq)
+
+initState :: Mock.EmptyValues vs  => BackingStoreState ks vs d
+initState = BackingStoreState {
+    bssMock  = Mock.emptyMock
+  , bssStats = initStats
+  }
+
+-- | Maximum number of LMDB readers that can be active at a time.
+--
+-- 32 is an arbitrary number of readers. We can increase or decrease this at
+-- will.
+--
+-- TODO: should the LMDB backing store keep track of the number of open value
+-- handles itself, and throw a custom error if the maximum is exceeded?
```

Comment:
It does, but it would be nicest if we do not have to match on different types of errors coming from different sources -- I've removed the TODO for now, since we don't test the case where the maximum is exceeded anyway